gems/aws-sdk-s3/lib/aws-sdk-s3/object_multipart_copier.rb

```diff
@@ -104,24 +109,33 @@ def compute_parts(size, default_part_size, options)
         parts = []
         options = options_for(:upload_part_copy, options)
         while offset < size
+          part_size = calculate_part_size(part_number, default_part_size, options)
           parts << options.merge({
             part_number: part_number,
-            copy_source_range: byte_range(offset, default_part_size, size),
+            copy_source_range: byte_range(offset, part_size, size),
           })
           part_number += 1
-          offset += default_part_size
+          offset += part_size
         end
         parts
       end
 
-      def byte_range(offset, default_part_size, size)
-        if offset + default_part_size < size
-          "bytes=#{offset}-#{offset + default_part_size - 1}"
+      def byte_range(offset, part_size, size)
+        if offset + part_size < size
+          "bytes=#{offset}-#{offset + part_size - 1}"
         else
           "bytes=#{offset}-#{size - 1}"
         end
       end
 
+      def calculate_part_size(part_number, default_part_size, options)
+        if @use_source_parts && source_metadata(options.merge({ part_number: 1 }))[:parts_count]
```

Comment:
using `source_metadata` here will make an additional call to `head_object` for each part we need to copy, but I believe this call should always return the same value.

Reply:
yes, this will always return the same value. I've extracted it to a memoized method so it only gets called once.